cpp/src/arrow/compute/kernels/scalar_nested_test.cc

```diff
@@ -39,6 +39,46 @@ TEST(TestScalarNested, ListValueLength) {
   }
 }
 
+TEST(TestScalarNested, ListElement) {
+  for (auto ty : {list(int16()), fixed_size_list(uint8(), 2)}) {
+    auto input = ArrayFromJSON(ty, "[[8, 5], [74, 92], [7, 4]]");
+    auto out_ty = ty->id() == Type::LIST ? int16() : uint8();
+    auto expected = ArrayFromJSON(out_ty, "[5, 92, 4]");
+    auto index = ScalarFromJSON(int32(), "1");
+    CheckVectorBinary("list_element", input, Datum(index), expected);
+  }
+
+  for (auto ty : {fixed_size_list(float32(), 2), large_list(float64())}) {
+    auto input = ArrayFromJSON(ty, "[[8.0, 5.0], [74.0, 92.0], [7.0, 4.0]]");
+    auto out_ty = ty->id() == Type::FIXED_SIZE_LIST ? float32() : float64();
+    auto expected = ArrayFromJSON(out_ty, "[8, 74.0, 7.0]");
+    auto index = ScalarFromJSON(int64(), "0");
+    CheckVectorBinary("list_element", input, Datum(index), expected);
+  }
+
+  // Construct a list with a non-empty null slot
+  auto input = ArrayFromJSON(list(int16()), "[[0, null, 1], [0, 0], [2, 3], [4, 5]]");
+  TweakValidityBit(input, 1, false);
+  auto expected = ArrayFromJSON(int16(), "[null, 0, 3, 5]");
+  auto index = ScalarFromJSON(int32(), "1");
+  CheckVectorBinary("list_element", input, Datum(index), expected);
+}
+
+TEST(TestScalarNested, ListElementChunkedArray) {
```

Comment:
CheckScalar will do this for you. It'll also check scalar inputs, which you may want to handle.

Reply:
By using `CheckScalar` I'm getting this:

`'_error_or_value12.status()' failed with NotImplemented: Function list_element has no kernel matching input types (scalar[list<item: int16>], scalar[int32])`

It seems `CheckScalar` is wrapping the type inside a scalar, do you know what I'm missing here @lidavidm ?

Reply:
cc @edponce

Reply:
Right, as stated, you'll have to support scalar inputs. CheckScalar will pull each individual element of each array as a scalar and call your function with those arguments too, to cover more input type shapes.

Reply:
You could do this by registering another kernel where the signature is `InputType::Scalar(Type::LIST), ...`. The implementation would cast to BaseListScalar (no need to template on list vs large list vs fixed-size list) and return basically `args[0].scalar_as<BaseListScalar>.value->GetScalar(index)`.

Reply:
This is the code that is throwing that issue
`CheckScalar("list_element", {input, Datum(index)}, expected);`
cc @edponce @lidavidm

Reply:
Thanks!!!